tests/Security/Http/EventListener/CheckTwoFactorCodeReuseListenerTest.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Tests\Security\Http\EventListener;
+
+use PHPUnit\Framework\Attributes\Test;
+use PHPUnit\Framework\MockObject\MockObject;
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Scheb\TwoFactorBundle\Security\Http\EventListener\CheckTwoFactorCodeListener;
+use Scheb\TwoFactorBundle\Security\Http\EventListener\CheckTwoFactorCodeReuseListener;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeCheckEvent;
+use Scheb\TwoFactorBundle\Security\TwoFactor\Event\TwoFactorCodeReusedEvent;
+use Scheb\TwoFactorBundle\Tests\TestCase;
+use stdClass;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\Security\Core\User\UserInterface;
+use function sha1;
+
+/**
+ * @property CheckTwoFactorCodeListener $listener
+ */
+class CheckTwoFactorCodeReuseListenerTest extends TestCase
+{
+    private MockObject|EventDispatcherInterface $eventDispatcher;
+
+    private MockObject|UserInterface $user;
+
+    private const CODE = '123456';
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->eventDispatcher = $this->createMock(EventDispatcherInterface::class);
+        $this->user = $this->createMock(UserInterface::class);
+        $this->user->method('getUserIdentifier')->willReturn('123');
+    }
+
+    protected function expectDoNothing(): void
+    {
+        $this->eventDispatcher
+            ->expects($this->never())
+            ->method($this->anything());
+    }
+
+    #[Test]
+    public function nothingHappensWithoutCacheProvider(): void
+    {
+        $listener = new CheckTwoFactorCodeReuseListener($this->eventDispatcher, null);
+        $this->expectDoNothing();
+
+        $listener->checkForCodeReuse(new TwoFactorCodeCheckEvent($this->user, self::CODE));
+    }
+
+    #[Test]
+    public function nothingHappensWithNoneCacheObjectAsCacheProvider(): void
+    {
+        $listener = new CheckTwoFactorCodeReuseListener($this->eventDispatcher, new stdClass());
+        $this->expectDoNothing();
+
+        $listener->checkForCodeReuse(new TwoFactorCodeCheckEvent($this->user, self::CODE));
+    }
+
+    #[Test]
+    public function nothingHappensWithPsrCacheProvider_noHit(): void
+    {
+        $cacheProvider = $this->createMock(CacheItemPoolInterface::class);
+        $cacheItem = new CacheItem();
+        $cacheProvider->expects($this->once())
+            ->method('getItem')->willReturnCallback(function (string $key) use ($cacheItem) {
+                $this->assertEquals('scheb_two_factor_code_reuse.'.sha1('123.123456'), $key);
+                $cacheItem->set(true);
+                $cacheItem->expiresAfter(60);
+
+                return $cacheItem;
+            });
```

Comment:
For ` 123` and `123456` descriptive constants should be used to make clear what these values represent. `123` would also need to be extracted into a `USER_IDENTIFIER` constant.

I'm also wondering, if the callback is really necessary. For testing the "no hit" behavior, it might even be sufficient to go with the same approach as below and just return a mocked CacheItem = `$cacheItem->expects($this->once())->method('isHit')->willReturn(false);`.

Reply:
Yeah. This tests in essence two different things:

* The `getItem` is called with the expected data 
* On the first invocation with a 2FA-code nothing happens as expected.

For that I need to make sure that the passed `$key` is actually the expected one. 

But I can do that similar to the `codeReusedEventIsDispatchedWithCacheHit` method using `with` and `return`.

Reply:
Simplified this test